src/MSBuildLocator/MSBuildLocator.Mono.cs

```diff
@@ -0,0 +1,147 @@
+                {
+                    return true;
+                }
+
+                // The path has a valid mono, but we can't find the version
+                // so, let's return the instance at least but with version 0.0.0
+                ver = new Version(0, 0, 0);
+                return true;
+            }
+
+            bool TryGetMonoVersionFromMonoBinary(string monoPrefixPath, out Version ver)
+            {
+                ver = null;
+                try
+                {
+                    using (var p = new Process())
+                    {
+                        p.StartInfo = new ProcessStartInfo
+                        {
+                            FileName = Path.Combine(monoPrefixPath, "bin", "mono"),
+                            Arguments = "--version=number",
+                            UseShellExecute = false,
+                            RedirectStandardOutput = true,
+                            RedirectStandardError = true
+                        };
+
+                        // Don't pollute caller's console
+                        p.OutputDataReceived += (s, e) => {};
+                        p.ErrorDataReceived += (s, e) => {};
+
+                        p.Start();
+                        p.WaitForExit();
+
+                        var stdout_str = p.StandardOutput.ReadToEnd();
+                        return Version.TryParse(stdout_str, out ver);
+                    }
+                }
+                catch (Win32Exception)
+                { }
+
+                return false;
+            }
+        }
+
+        // Taken from MSBuild/NativeMethodsShared
+        private static readonly object IsRunningOnMonoLock = new object();
+
+        private static bool? _isRunningOnMono;
+
+        /// <summary>
+        /// Gets a flag indicating if we are running under MONO
+        /// </summary>
+        internal static bool IsRunningOnMono
+        {
+            get
+            {
+                if (_isRunningOnMono.HasValue) return _isRunningOnMono.Value;
+
+                lock (IsRunningOnMonoLock)
+                {
+                    if (_isRunningOnMono == null)
+                    {
+                        // There could be potentially expensive TypeResolve events, so cache IsMono.
+                        _isRunningOnMono = Type.GetType("Mono.Runtime") != null;
+                    }
+                }
+
+                return _isRunningOnMono.Value;
+            }
+        }
+
+        // Taken from MSBuild/NativeMethodsShared
+        private static bool? _isOSX;
+
+        /// <summary>
+        /// Gets a flag indicating if we are running under Mac OSX
+        /// </summary>
+        internal static bool IsOSX
+        {
+            get
+            {
+                if (!_isOSX.HasValue)
+                {
+                    _isOSX = File.Exists("/usr/lib/libc.dylib");
```

Comment:
Doesn't this break on Big Sur? Can you call `RuntimeInformation.IsOSPlatform` instead? And in that case it might be worth just inlining the call instead of caching it here.